flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -325,8 +366,12 @@ private void stopInternalServices() {
 
     public CompletableFuture<Void> stop(StopMode mode) {
         switch (operatingMode) {
+            case WAITING_FOR_ALIGNMENT:
             case OUTPUT_NOT_INITIALIZED:
             case READING:
+                if (operatingMode == OperatingMode.WAITING_FOR_ALIGNMENT) {
+                    waitingForAlignmentFuture.complete(null);
```

Comment:
Don't know if better, but can't we instead combine the `waitingForAlignmentFuture` with `finished` in `getAvailableFuture`?

Something like:
```
    public CompletableFuture<?> getAvailableFuture() {
....
            case WAITING_FOR_ALIGNMENT:
                return availabilityHelper.update(waitingForAlignmentFuture);
            case OUTPUT_NOT_INITIALIZED:
            case READING:
                return availabilityHelper.update(sourceReader.isAvailable());
```

Reply:
I was thinking about it, initially I decided against it, but on the second thought maybe indeed that's better.